software/examples/mqtt/example-configuration.txt
# Change XXYYZZ to the UID of your Servo Brick

setup:
	# Configure two servos with voltage 5.5V
	# Servo 1: Connected to port 0, period of 19.5ms, pulse width of 1 to 2ms
	#          and operating angle -100 to 100°
	#
	# Servo 2: Connected to port 5, period of 20ms, pulse width of 0.95
	#          to 1.95ms and operating angle -90 to 90°
	publish '{"voltage": 5500}' to tinkerforge/request/servo_brick/XXYYZZ/set_output_voltage

	publish '{"servo_num": 0, "min": -10000, "max": 10000}' to tinkerforge/request/servo_brick/XXYYZZ/set_degree

	publish '{"servo_num": 0, "min": 1000, "max": 2000}' to tinkerforge/request/servo_brick/XXYYZZ/set_pulse_width

	publish '{"servo_num": 0, "period": 19500}' to tinkerforge/request/servo_brick/XXYYZZ/set_period

	publish '{"servo_num": 0, "acceleration": 1000}' to tinkerforge/request/servo_brick/XXYYZZ/set_acceleration # Slow acceleration

	publish '{"servo_num": 0, "velocity": 65535}' to tinkerforge/request/servo_brick/XXYYZZ/set_velocity # Full speed

	publish '{"servo_num": 5, "min": -9000, "max": 9000}' to tinkerforge/request/servo_brick/XXYYZZ/set_degree

	publish '{"servo_num": 5, "min": 950, "max": 1950}' to tinkerforge/request/servo_brick/XXYYZZ/set_pulse_width

	publish '{"servo_num": 5, "period": 20000}' to tinkerforge/request/servo_brick/XXYYZZ/set_period

	publish '{"servo_num": 5, "acceleration": 65535}' to tinkerforge/request/servo_brick/XXYYZZ/set_acceleration # Full acceleration

	publish '{"servo_num": 5, "velocity": 65535}' to tinkerforge/request/servo_brick/XXYYZZ/set_velocity # Full speed

	publish '{"servo_num": 0, "position": 10000}' to tinkerforge/request/servo_brick/XXYYZZ/set_position # Set to most right position

	publish '{"servo_num": 0}' to tinkerforge/request/servo_brick/XXYYZZ/enable

	publish '{"servo_num": 5, "position": -9000}' to tinkerforge/request/servo_brick/XXYYZZ/set_position # Set to most left position

	publish '{"servo_num": 5}' to tinkerforge/request/servo_brick/XXYYZZ/enable

cleanup:
	# If you are done, run this to clean up
	publish '{"servo_num": 0}' to tinkerforge/request/servo_brick/XXYYZZ/disable

	publish '{"servo_num": 5}' to tinkerforge/request/servo_brick/XXYYZZ/disable


software/examples/mathematica/ExampleConfiguration.nb.txt
Needs["NETLink`"]
LoadNETAssembly["Tinkerforge",NotebookDirectory[]<>"../../.."]

host="localhost"
port=4223
uid="XXYYZZ"(*Change XXYYZZ to the UID of your Servo Brick*)

(*Create IPConnection and device object*)
ipcon=NETNew["Tinkerforge.IPConnection"]
servo=NETNew["Tinkerforge.BrickServo",uid,ipcon]
ipcon@Connect[host,port]

(*Configure two servos with voltage 5.5V*)
(*Servo 1: Connected to port 0, period of 19.5ms, pulse width of 1 to 2ms*)
(*         and operating angle -100 to 100°*)
(*Servo 2: Connected to port 5, period of 20ms, pulse width of 0.95 *)
(*         to 1.95ms and operating angle -90 to 90°*)
servo@SetOutputVoltage[5500]

servo@SetDegree[0,-10000,10000]
servo@SetPulseWidth[0,1000,2000]
servo@SetPeriod[0,19500]
servo@SetAcceleration[0,1000](*Slow acceleration*)
servo@SetVelocity[0,65535](*Full speed*)

servo@SetDegree[5,-9000,9000]
servo@SetPulseWidth[5,950,1950]
servo@SetPeriod[5,20000]
servo@SetAcceleration[5,65535](*Full acceleration*)
servo@SetVelocity[5,65535](*Full speed*)

servo@SetPosition[0,10000](*Set to most right position*)
servo@Enable[0]

servo@SetPosition[5,-9000](*Set to most left position*)
servo@Enable[5]

Input["Click OK to exit"]

servo@Disable[0]
servo@Disable[5]

ipcon@Disconnect[]
ReleaseNETObject[servo]
ReleaseNETObject[ipcon]


software/changelog
2011-11-24: 1.0.0 (cc2f075)
- Initial version

2012-02-15: 1.1.0 (b80a126)
- Add support for tick types

2012-03-05: 1.1.1 (0b80822)
- Fix bug with timeouts on Windows (through bricklib)

2012-03-11: 1.1.2 (a7235ec)
- Fix bug with ADC calibration readout (through bricklib)

2012-06-30: 1.1.3 (91bfc80)
- Add get_chip_temperature and reset function

2012-07-02: 1.1.4 (c1bcd00)
- Add support for minimum period of 1us and add USB hotplug support

2012-10-26: 1.1.5 (bbcad00)
- Change Bricklet Context size from 250 to 256 (through bricklib)

2013-01-17: 2.0.0 (3c4227a)
- Support for Protocol 2.0

2013-03-27: 2.0.1 (2e5ca12)
- Add enable/disable API for reached callbacks

2013-04-11: 2.0.2 (5a58710)
- Add watchdog timer (restart after 16s of "deadlock")

2013-06-17: 2.0.3 (9eb0db7)
- Add missing defines for enable/disable API for reached callbacks

2013-06-27: 2.0.4 (d53add1)
- Use tick_task to detect USB connect/disconnect instead of interrupt

2014-07-07: 2.0.5 (77e3ed8)
- Remove USB hotplug support (through bricklib)

2014-11-25: 2.3.0 (1659022)
- Add support for new SPI protocol

2016-08-29: 2.3.1 (fd8c76a)
- Make sure to use all chunks when writing Bricklet plugin (through bricklib)

2017-05-11: 2.3.2 (5903619)
- Add support for Co-MCU Bricklets

2017-06-28: 2.3.3 (e737cdf)
- Fix incompatibility between Co-MCU support and LED Strip Bricklet (through bricklib)

2017-12-13: 2.3.4 (7e884e7)
- Make sure Brick does not restart on suspend/resume (through bricklib)
- Refactor initial enumeration handling, fix double-enumeration bugs (through bricklib)

2018-01-11: 2.3.5 (83558bd)
- Fix message duplication bug (through bricklib)

2018-04-27: 2.3.6 (15993f3)
- Fix potential deadlock between USB stack and Co-MCU enumeration (through bricklib)

2018-07-31: 2.3.7 (28e9cd9)
- Add support for Isolator Bricklet (through bricklib)
- Improve Co-MCU Bricklet error handling (through bricklib)

2019-02-27: 2.3.8 (c7067a8)
- Changes through bricklib:
  - Fix Bricklet port miss-identification
  - Increase SPI stack timeout (for RED Brick)
  - Fix bugs in set/get_spitfp_* functions

2019-11-22: 2.3.9 (ad70b6a):
- Make sure that Servo Brick output goes low if pulse width and position is set to 0.


software/src/config.h
/* servo-brick
 * Copyright (C) 2010-2013 Olaf Lüke <olaf@tinkerforge.com>
 *
 * config.h: Servo-Brick specific configuration
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef CONFIG_H
#define CONFIG_H

#include "servo.h"
#include "communication.h"
#include "bricklib/drivers/board/sam3s/SAM3S.h"

#define BRICK_FIRMWARE_VERSION_MAJOR 2
#define BRICK_FIRMWARE_VERSION_MINOR 3
#define BRICK_FIRMWARE_VERSION_REVISION 9

#define BRICK_HARDWARE_VERSION_MAJOR 1
#define BRICK_HARDWARE_VERSION_MINOR 0
#define BRICK_HARDWARE_VERSION_REVISION 0

#define BRICK_DEVICE_IDENTIFIER 14

// ************** DEBUG SETTINGS **************
#define DEBUG_SPI_STACK 0
#define DEBUG_I2C_EEPROM 0
#define DEBUG_STARTUP 1
#define DEBUG_BRICKLET 1
#define DEBUG_SERVO 1
//#define PROFILING
//#define PROFILING_TIME 100 // After how many seconds profiling is printed

#define DISABLE_JTAG_ON_STARTUP
//#define LOGGING_SERIAL
//#define LOGGING_LEVEL LOGGING_DEBUG
#define LOGGING_LEVEL LOGGING_NONE


// ************** BRICK SETTINGS **************

// Frequencies
#define BOARD_MCK      64000000 // Frequency of brick
#define BOARD_MAINOSC  16000000 // Frequency of oscillator
#define BOARD_ADC_FREQ 16000000 // Frequency of ADC
#define BOARD_OSC_EXTERNAL      // Use external oscillator

#define BRICK_HAS_CO_MCU_SUPPORT

// UART for console output (printf)
#define PIN_CONSOLE_RXD  {1 << 21, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_CONSOLE_TXD  {1 << 22, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

#define CONSOLE_BAUDRATE 115200
#define CONSOLE_USART    USART1
#define CONSOLE_ID       ID_USART1
#define CONSOLE_PINS     {PIN_CONSOLE_RXD, PIN_CONSOLE_TXD}

#define PINS_UART        {PIN_CONSOLE_RXD, PIN_CONSOLE_TXD}

// TWI
// TWI version
#define TWI_V3XX

// TWI stack definitions (for reading of eeproms from Bricks in stack)
#define TWI_STACK           TWI1
#define PIN_TWI_TWD_STACK   {1 << 4, PIOB, ID_PIOB, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_TWI_TWCK_STACK  {1 << 5, PIOB, ID_PIOB, PIO_PERIPH_A, PIO_DEFAULT}
#define PINS_TWI_STACK      PIN_TWI_TWD_STACK, PIN_TWI_TWCK_STACK

// TWI bricklet definitions (for bricklets, spi select and brick specific
//                           functions)
#define TWI_BRICKLET           TWI0
#define PIN_TWI_TWD_BRICKLET   {1 << 3, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_TWI_TWCK_BRICKLET  {1 << 4, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PINS_TWI_BRICKLET      PIN_TWI_TWD_BRICKLET, PIN_TWI_TWCK_BRICKLET

// USB
// USB VBUS monitoring pin for USB plug and play
#define PIN_USB_DETECT  {1 << 28, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
// USB product descriptor (name of brick)
#define PRODUCT_DESCRIPTOR { \
	USBStringDescriptor_LENGTH(11), \
    USBGenericDescriptor_STRING, \
    USBStringDescriptor_UNICODE('S'), \
    USBStringDescriptor_UNICODE('e'), \
    USBStringDescriptor_UNICODE('r'), \
    USBStringDescriptor_UNICODE('v'), \
    USBStringDescriptor_UNICODE('o'), \
    USBStringDescriptor_UNICODE(' '), \
    USBStringDescriptor_UNICODE('B'), \
    USBStringDescriptor_UNICODE('r'), \
    USBStringDescriptor_UNICODE('i'), \
    USBStringDescriptor_UNICODE('c'), \
    USBStringDescriptor_UNICODE('k') \
}


// SPI
#define PIN_SPI_MISO        {1 << 12, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_SPI_MOSI        {1 << 13, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_SPI_SPCK        {1 << 14, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

#define PINS_SPI            PIN_SPI_MISO, PIN_SPI_MOSI, PIN_SPI_SPCK

#define PIN_SPI_SELECT_SLAVE  {1 << 11, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

// LED
#define PIN_LED_STD_BLUE    {1 << 9, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_LED_STD_RED     {1 << 10, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}
#define PINS_STD_LED        PIN_LED_STD_BLUE, PIN_LED_STD_RED
#define PINS_LED            PIN_LED_STD_BLUE, PIN_LED_STD_RED

#define LED_STD_BLUE        0
#define LED_STD_RED         1


// Brick Detect
// Set low by master
// TODO: Change name
#define PIN_DETECT        {1 << 6, PIOB, ID_PIOB, PIO_OUTPUT_0, PIO_DEFAULT}



// ************** INTERRUPT PRIORITIES ***********
#define PRIORITY_SERVO_TC0           4
#define PRIORITY_MCP3008_USART0      6
#define PRIORITY_PERIODIC_ADC_TC2   15
#define PRIORITY_EEPROM_MASTER_TWI0  1
#define PRIORITY_EEPROM_SLAVE_TWI1   1
#define PRIORITY_STACK_SLAVE_SPI     5
#define PRIORITY_PROFILING_TC0       0

// ************** BRICKLET SETTINGS **************

// Number of bricklet ports
#define BRICKLET_NUM 2

// BRICKLET A
#define BRICKLET_A_ADDRESS 84
#define BRICKLET_A_ADC_CHANNEL 4

#define BRICKLET_A_PIN_1_AD   {1 << 0, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_2_DA   {1 << 13, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_3_PWM  {1 << 15, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_4_IO   {1 << 29, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_SELECT {0, 0, 0, 0, 0}

// BRICKLET B
#define BRICKLET_B_ADDRESS 80
#define BRICKLET_B_ADC_CHANNEL 5

#define BRICKLET_B_PIN_1_AD   {1 << 1, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_2_DA   {1 << 3, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_3_PWM  {1 << 30, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_4_IO   {1 << 1, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_SELECT {0, 0, 0, 0, 0}


// ************** SERVO SETTINGS **************
#define SERVO_PWM_PIN_0 {1 << 23, PIOA, ID_PIOA, PIO_PERIPH_B, PIO_DEFAULT}
#define SERVO_PWM_PIN_1 {1 << 17, PIOA, ID_PIOA, PIO_PERIPH_C, PIO_DEFAULT}
#define SERVO_TC_PIN_2  {1 << 16, PIOA, ID_PIOA, PIO_PERIPH_B, PIO_DEFAULT}
#define SERVO_PWM_PIN_3 {1 << 24, PIOA, ID_PIOA, PIO_PERIPH_B, PIO_DEFAULT}
#define SERVO_PWM_PIN_4 {1 << 25, PIOA, ID_PIOA, PIO_PERIPH_B, PIO_DEFAULT}
#define SERVO_TC_PIN_5  {1 << 26, PIOA, ID_PIOA, PIO_PERIPH_B, PIO_DEFAULT}
#define SERVO_TC_PIN_6  {1 << 0,  PIOA, ID_PIOA, PIO_PERIPH_B, PIO_DEFAULT}

#define SERVO_OFF_PIN_0 {1 << 23, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define SERVO_OFF_PIN_1 {1 << 17, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define SERVO_OFF_PIN_2 {1 << 16, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define SERVO_OFF_PIN_3 {1 << 24, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define SERVO_OFF_PIN_4 {1 << 25, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define SERVO_OFF_PIN_5 {1 << 26, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define SERVO_OFF_PIN_6 {1 << 0,  PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}


#define SERVO_ON_PINS  SERVO_PWM_PIN_0, SERVO_PWM_PIN_1, SERVO_TC_PIN_2, \
                       SERVO_PWM_PIN_3, SERVO_PWM_PIN_4, SERVO_TC_PIN_5, \
                       SERVO_TC_PIN_6

#define SERVO_OFF_PINS SERVO_OFF_PIN_0, SERVO_OFF_PIN_1, SERVO_OFF_PIN_2, \
                       SERVO_OFF_PIN_3, SERVO_OFF_PIN_4, SERVO_OFF_PIN_5, \
                       SERVO_OFF_PIN_6

// ************** SERVO CURRENT MEASARING *****
#define USART_SPI_CLK  1000000 // 1 Mhz

#define PIN_USART_SPI_MISO          {1 << 5, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_USART_SPI_MOSI          {1 << 6, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_USART_SPI_SPCK          {1 << 2, PIOA, ID_PIOA, PIO_PERIPH_B, PIO_DEFAULT}
#define PIN_USART_SPI_SELECT_SLAVE  {1 << 7, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

#define PINS_USART_SPI   PIN_USART_SPI_MISO, PIN_USART_SPI_MOSI, PIN_USART_SPI_SPCK, PIN_USART_SPI_SELECT_SLAVE

// ************** SERVO DEBUGGING **************
#if(DEBUG_SERVO)
#define logservod(str, ...) do{logd("servo: " str, ##__VA_ARGS__);}while(0)
#define logservoi(str, ...) do{logi("servo: " str, ##__VA_ARGS__);}while(0)
#define logservow(str, ...) do{logw("servo: " str, ##__VA_ARGS__);}while(0)
#define logservoe(str, ...) do{loge("servo: " str, ##__VA_ARGS__);}while(0)
#define logservof(str, ...) do{logf("servo: " str, ##__VA_ARGS__);}while(0)
#else
#define logservod(str, ...) {}
#define logservoi(str, ...) {}
#define logservow(str, ...) {}
#define logservoe(str, ...) {}
#define logservof(str, ...) {}
#endif
#endif


// ************** POWER MANAGEMENT **************
#define SERVO_DEFAULT_OUTPUT_VOLTAGE 5000
#define SERVO_OUTPUT_VOLTAGE_PIN  {1 << 14, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define SERVO_OUTPUT_VOLTAGE_CHANNEL 1
#define SERVO_POWER_ENABLE_PIN  {1 << 20, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}

#define VOLTAGE_MAX_VALUE 4095

#define VOLTAGE_STACK_CHANNEL 2
#define VOLTAGE_STACK_MULTIPLIER 11
#define VOLTAGE_STACK_REFERENCE 3300
#define VOLTAGE_STACK_PIN {1 << 19, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define VOLTAGE_STACK_SWITCH_PIN {1 << 27, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}

#define VOLTAGE_EXTERN_CHANNEL 1
#define VOLTAGE_EXTERN_MULTIPLIER 11
#define VOLTAGE_EXTERN_REFERENCE 3300
#define VOLTAGE_EXTERN_PIN {1 << 18, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}

